pkgs/tools/nix/nixos-render-docs/README.md

```diff
@@ -0,0 +1,82 @@
+
+Several alternatives were considered, a detailed analysis on the trade-offs with each alternative can be found here: https://ethercalc.net/dc4vcnnl8zv0
+
+## Redirects system
+
+Moving contents around can cause links to break.
+
+Since we have our own markdown parser, we can hook into the rendering process to extract all of the metadata around each content identifier.
+The [mechanism for checking correctness of redirects](./src/nixos_render_docs/redirects.py) takes the collection of identifiers and a static mapping of redirects, validates them against a set of rules, and creates a client-side redirect mapping for each output path, as well as a `_redirects` file for server-side redirects in [Netlify syntax](https://docs.netlify.com/routing/redirects/#syntax-for-the-redirects-file).
+
+This allows us to catch:
+- Identifiers that were removed or renamed
+- Content that was moved from one location to another
+- Various consistency errors in the redirects mapping
+
+### Design considerations
+
+The creation, movement, and removal of every identifier is captured in the Git history of the Nixpkgs repo.
+However, leveraging this information during the build process of `nixos-render-docs` is quite tricky.
+There are two key inputs to the redirect system:
+
+1. The current set of identifiers
+2. A list of their historical locations
+
+One is accessible during the build process, and the other by fetching the Git history.
+Doing both of these during the build is not feasible, and to have an external command handling the Git history would require some form of communication between it and the build process.
+This quickly gets complicated.
+
+The chosen design is a trade-off between speed, repository size, and contributor friction:
+- Stricter checks always require more attention from contributors
+    - Checks should be reasonably fast and ideally happen locally, e.g. as part of the build, as anything else will substantially lengthen the feedback cycle.
+- Computing redirects against previous revisions of the repository would be more space-efficient, but impractically slow.
+    - It would also require keeping an impure or ever-moving reference to those other revisions.
+    - The static mapping acts like a semi-automatically updated cache that we drag along with version history
+    - Other setups, such as a dedicated service to cache a history of moved content, are more complicated and would still be impure
+- Checking in large amounts of data that is touched often, bears a risk of more merge conflicts or related build failures
+
+
+A solution picked here is to have a static mapping of the historical locations checked into the Git tree, such that it can be read during the build process.
```

Comment:
```suggestion
The solution picked here is to have a static mapping of the historical locations checked into the Git tree, such that it can be read during the build process.
```